For Google Host Review

shaniyaclement opened the pull request:
<!--
 Licensed to the Apache Software Foundation (ASF) under one
 or more contributor license agreements.  See the NOTICE file
 distributed with this work for additional information
 regarding copyright ownership.  The ASF licenses this file
 to you under the Apache License, Version 2.0 (the
 "License"); you may not use this file except in compliance
 with the License.  You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing,
 software distributed under the License is distributed on an
 "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 KIND, either express or implied.  See the License for the
 specific language governing permissions and limitations
 under the License.
 -->

<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of an existing issue, reference it using one of the following:

closes: #ISSUE
related: #ISSUE

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

Commit:
CreateDataPipeline operator and hook created.

Commit:
Added DAG for datapipeline to test if running as hoped

Commit:
created the run pipeline hook, updated the data pipeline create hook and operator to properly intake the body

Commit:
updated the create pipeline to accept the body. created run pipeline hook. updated operator accordingly

Commit:
Add RunDataPipelineOperator

blpham committed:
Merge pull request #1 from blpham/runop

Add RunDataPipelineOperator

Commit:
Example dags is working for creating a new pipeline. Committing minor chagnes before working on run operator/ hook exanple dag and testing

Commit:
return full response body update

Commit:
The Create and Run operators are meant to work together, so it was recommended that the run operator accept the entire response body.
Updated the Run Operator to accept the response body dict and removed the other parameters because they were used to build the pipeline name, which the response body includes.
Created variable for the pipeline name that takes it from the given response body. Used it to call upon the hook.
Updated the hook to accept the pipeline name and removed the building of the name.
Added run_data_pipeline to example dags to call the Run Data Pipeline operator passing a pipeline name taken from the create_data_pipeline result. (which should be a response dict).

Commit:
Update variable names to be consistent

shaniyaclement committed:
Merge pull request #2 from blpham/update1

Update variable names to be consistent

Commit:
Update the Run Operator to accept  the Pipeline name instead of the response body.

Commit:
Update example dags to reflect that only the name is being returned from the Create Operator

Commit:
Create Operator returns full response body. example dag reflects that

blpham committed:
Merge branch 'main' into Shaniya's-Commits

blpham committed:
Merge pull request #4 from shaniyaclement/Shaniya's-Commits

Update Create and Run operator to run and accept the data pipeline name

Commit:
fixed syntax errors. gave run example dag a task id and pipeline name for testing

Commit:
Added project_id and location to run pipeline hook.
Updated the dags to work with dataflow-interns. Running properly.
Added project_id and location to run operator.

shaniyaclement committed:
Merge pull request #5 from shaniyaclement/main

Updating Commits branch to match main updates

Commit:
add testing file

shaniyaclement committed:
Merge pull request #6 from shaniyaclement/Shaniya's-Commits

Adding testing file

Commit:
Change created pipeline source to airflow

Commit:
Change created pipeline source to airflow

Commit:
Change created pipeline source to airflow

Commit:
Change created pipeline source to airflow

Commit:
Change created pipeline source to airflow

Commit:
Change created pipeline source to airflow

Commit:
Fix source format and remove whitespace

Commit:
Fix source format and remove whitespace

Commit:
Fix source format and remove whitespace

Commit:
Fix source format and remove whitespace

Commit:
Fix source format and remove whitespace

shaniyaclement committed:
Merge pull request #7 from blpham/update2

Change created pipeline source to airflow in CreateDataPipelineOperator

Commit:
Raise exception if data_pipeline_name not given

Commit:
Raise exception if data_pipeline_name not given

Commit:
Raise exception if data_pipeline_name not given

Commit:
Raise exception if data_pipeline_name not given

Commit:
Raise exception if data_pipeline_name not given

Commit:
Raise exception if data_pipeline_name not given

Commit:
Removed unnecessary variable data_pipeline_name from operator and all placed referenced for the create.
Added code comments to explain parameters for operators and hooks.
Updated run hook to utilize all variables.
Removed some whitespace.

Commit:
Updated code comments

blpham committed:
Merge pull request #8 from shaniyaclement/Shaniya's-Commits

Code Comments, White Space, Variable usage

Commit:
Updated comments to include that the return values

Commit:
Create hooks test file, rename operator test file

Commit:
Create hooks test file, rename operator test file

Commit:
Create hooks test file, rename operator test file

Commit:
Create hooks test file, rename operator test file

Commit:
Create hooks test file, rename operator test file

Commit:
Create hooks test file, rename operator test file

Commit:
Create hooks test file, rename operator test file

shaniyaclement committed:
Merge pull request #9 from blpham/update3

Create file for testing hooks, rename operator test file

shaniyaclement committed:
Merge branch 'apache:main' into Shaniya's-Commits

shaniyaclement committed:
Merge branch 'apache:main' into main

Commit:
Updating Branch

Commit:
Add Airflow Exceptions for run operator

Commit:
Add Airflow Exceptions for run operator

shaniyaclement committed:
Merge pull request #12 from blpham/update4

Add Airflow Exceptions for run operator

Commit:
Test variables and operator used for testing create op. Some imports that I  believe will be necessary for testing create op

Commit:
Merge Conflicts resolved

shaniyaclement committed:
Merge branch 'main' into Shaniya's-Commits

shaniyaclement committed:
Merge pull request #10 from shaniyaclement/Shaniya's-Commits

Updating Branches to Match

Commit:
Added Airflow exception for if body parameter is not given. Cannot create the pipeline without it.

Commit:
Fixed indentation

xianhualiu commented on airflow/providers/google/cloud/hooks/datapipeline.py:
```diff
@@ -0,0 +1,129 @@
+        
+        parent = self.build_parent_name(project_id, location)
+        service = self.get_conn()
+        print(dir(service.projects().locations()))
```
replace print with self.log.info

shaniyaclement replied:
done

xianhualiu commented on airflow/providers/google/cloud/hooks/datapipeline.py:
```diff
@@ -0,0 +1,129 @@
+        """
+        parent = self.build_parent_name(project_id, location)
+        service = self.get_conn()
+        print(dir(service.projects().locations()))
```
replace print with self.log.info

shaniyaclement replied:
done

xianhualiu commented on airflow/providers/google/cloud/operators/datapipeline.py:
```diff
@@ -0,0 +1,139 @@
+            location = self.location,
+        )
+        self.log.info("Response Body: ", self.data_pipeline)
+
```
please add failure handle logic based on the response code. If the response contains "error", AirflowException needs to be raised with the "message" of the "error".

shaniyaclement replied:
done

xianhualiu commented on airflow/providers/google/cloud/operators/datapipeline.py:
```diff
@@ -0,0 +1,139 @@
+            project_id = self.project_id,
+            location = self.location,
+        )
+
```
please add failure handle logic based on the response code. If the response contains "error", AirflowException needs to be raised with the "message" of the "error".

xianhualiu commented on tests/providers/google/cloud/operators/test_datapipeline.py:
```diff
@@ -0,0 +1,72 @@
+                    "tempLocation": "gs://intern-bucket-1/temp"
+                    },
+                    "parameters": {
+                    "inputFile": "gs://intern-bucket-1/examples/kinglear.txt",
```
please use more permanent gs pathes like https://github.com/apache/airflow/blob/375d2fabcc2cef4b17d2129b557ac6fcb6c65067/tests/providers/google/cloud/operators/test_dataflow.py#L42C3-L42C3

For example:
   ""containerSpecGcsPath": "gs://dataflow-templates-us-central1/latest/Word_Count_metadata"
    "inputFile": "gs://dataflow-samples/shakespeare/kinglear.txt",
    "output": "gs://test/output/my_output",

shaniyaclement replied:
I was required to push this code to create the PR, but the testing file is not complete. I included a TODO to reference needing to change the body pathes. Sorry, I wasn't aware you were going to review the testing file because you said the functions

Commit:
address PR comments

Commit:
accidentally deleted the service access line

Commit:
Addresses PR comment

manavgarg commented on airflow/providers/google/cloud/hooks/datapipeline.py:
```diff
@@ -0,0 +1,129 @@
+
+class DataPipelineHook(GoogleBaseHook):
+    """
+    Hook for Google DataPipeline.
```
nit (reword): Hook for Google Cloud Data Pipelines.

manavgarg commented on airflow/providers/google/cloud/hooks/datapipeline.py:
```diff
@@ -0,0 +1,129 @@
+        )
+
+    def get_conn(self) -> build:
+        """Returns a Google Cloud DataPipeline service object."""
```
nit: change DataPipeline to Data Pipelines

manavgarg commented on airflow/providers/google/cloud/hooks/datapipeline.py:
```diff
@@ -0,0 +1,129 @@
+        location: str = DEFAULT_DATAPIPELINE_LOCATION,
+    ) -> None:
+        """
+        Creates a new Data Pipeline instance from the Data Pipeline API.
```
nit: Pipeline -> Pipelines

manavgarg commented on airflow/providers/google/cloud/hooks/datapipeline.py:
```diff
@@ -0,0 +1,129 @@
+        :param body: The request body (contains instance of Pipeline). See:
+            https://cloud.google.com/dataflow/docs/reference/data-pipelines/rest/v1/projects.locations.pipelines/create#request-body
+        :param project_id: The ID of the GCP project that owns the job.
+        :param location: The location to direct the Data Pipeline instance to (example_dags uses uscentral-1).
```
nit: Pipeline -> Pipelines

manavgarg commented on airflow/providers/google/cloud/hooks/datapipeline.py:
```diff
@@ -0,0 +1,129 @@
+        location: str = DEFAULT_DATAPIPELINE_LOCATION,
+    ) -> None:
+        """
+        Runs a Data Pipeline Instance using the Data Pipeline API 
```
nit: Pipeline -> Pipelines

manavgarg commented on airflow/providers/google/cloud/hooks/datapipeline.py:
```diff
@@ -0,0 +1,129 @@
+        :param data_pipeline_name:  The display name of the pipeline. In example 
+            projects/PROJECT_ID/locations/LOCATION_ID/pipelines/PIPELINE_ID it would be the PIPELINE_ID.
+        :param project_id: The ID of the GCP project that owns the job.
+        :param location: The location of the Data Pipeline instance to (example_dags uses uscentral-1).
```
nit: Pipeline -> Pipelines

manavgarg commented on airflow/providers/google/cloud/operators/datapipeline.py:
```diff
@@ -0,0 +1,139 @@
+
+class CreateDataPipelineOperator(GoogleCloudBaseOperator):
+    """ 
+    Creates a new Data Pipeline instance from the Data Pipeline API.
```
nit: Pipeline -> Pipelines

manavgarg replied:
change this elsewhere in the file where applicable.

manavgarg commented on airflow/providers/google/cloud/operators/datapipeline.py:
```diff
@@ -0,0 +1,139 @@
+    :param gcp_conn_id: The connection ID to connect to the Google Cloud
+        Platform.
+
+    Returns the created Pipeline instance in JSON representation.
```
nit: Pipeline -> Data Pipelines

shaniyaclement committed:
Merge pull request #14 from shaniyaclement/Shaniya's-Commits

Testing File Updates, Address PR comments

manavgarg commented on airflow/providers/google/cloud/operators/datapipeline.py:
```diff
@@ -0,0 +1,139 @@
+        )
+        self.log.info("Response Body: ", self.data_pipeline)
+
+        # returns the full response body
```
remove the comment.

manavgarg commented on airflow/providers/google/cloud/operators/datapipeline.py:
```diff
@@ -0,0 +1,139 @@
+            body = self.body,
+            location = self.location,
+        )
+        self.log.info("Response Body: ", self.data_pipeline)
```
do we need to actually log the response body?

manavgarg commented on airflow/providers/google/cloud/operators/datapipeline.py:
```diff
@@ -0,0 +1,139 @@
+
+class RunDataPipelineOperator(GoogleCloudBaseOperator):
+    """ 
+    Runs a Data Pipeline Instance using the Data Pipeline API 
```
nit: Pipeline -> Pipelines

shaniyaclement committed:
Merge pull request #16 from shaniyaclement/main

Updating Working Branch with Main

manavgarg reviewed (commented):
Have posted some minor comments. Will take another look once the tests are ready.
Also, please change the PR Title to something like "Adding Support for Google Cloud's Data Pipelines Operators". And please add a summary of the key features in the PR's description.

Commit:
Re: xianhualiu; Raise exception if pipeline run error

Commit:
Re: xianhualiu; Raise exception if pipeline run error

Commit:
Passing Test Cases if the Create DataPipeline Operator is given the correct input

Commit:
Passing test for if projectid param is not given.

Commit:
Passing tests that check that the operator is passed the body and location.

Commit:
Create test for run operator execute

Commit:
Create test for run operator execute

Commit:
Accidentally added to wrong branch, undoing

Commit:
undoing

Commit:
undoing

Commit:
Passing test that if the response body contains an error an Airflow Exception is raised

shaniyaclement committed:
Merge pull request #17 from blpham/update8

Re: xianhualiu; Raise exception if pipeline run error

Commit:
Removed the Run Operator test because everything was not pushed

blpham committed:
Merge branch 'main' into Shaniya's-Commits

shaniyaclement committed:
Merge pull request #20 from shaniyaclement/Shaniya's-Commits

Create Data Pipeline Test Cases

shaniyaclement committed:
Merge pull request #21 from shaniyaclement/main

Updating Shaniya's Commits to match Main

Commit:
Create tests for run operator

Commit:
Remove import; using mock for hook

Commit:
Change test names to be consistent with create operator tests

Commit:
Commit for host to review. Renders an error that gcp_conn_id doesn't exist bc it is not mocking properly and that .create is not called in reference to line 127. these should only get feedback and not be approved to merge

Commit:
.

Commit:
minor fix to params in mock request

Commit:
Create hooks tests for running pipeline

Commit:
Update grammar

Commit:
Remove comment

Commit:
update

Commit:
Update

shaniyaclement committed:
Merge pull request #22 from blpham/update9

Add tests for run operator

shaniyaclement committed:
Revert "Add tests for run operator"

shaniyaclement committed:
Merge pull request #24 from blpham/update11

Re: manavgarg, address PR comments on wording

shaniyaclement committed:
Merge pull request #25 from shaniyaclement/revert-22-update9

Revert "Add tests for run operator"

shaniyaclement committed:
Merge pull request #26 from shaniyaclement/main

Updating Commits Branch to Match Main

Commit:
Minor removals based on Manav comments

blpham committed:
Revert "Revert "Add tests for run operator""

Commit:
Test hook for running data pipeline

blpham committed:
Merge branch 'shaniyaclement:main' into hook_tests

Commit:
remove comment

Commit:
test_create_data_pipeline was failing because "create()"
was not being called. this was because I used the wrong test_parent and my project id wasn't the same all around. fixed this and removed unnecessary patch. now focusing on mock gcp conn issue.

Commit:
Assert that run_data_pipeline call returns job

blpham committed:
Merge pull request #27 from shaniyaclement/revert-25-revert-22-update9

Add tests for run operator

Commit:
completed code. cleaned up unnecessary imports

shaniyaclement committed:
Merge branch 'main' into Shaniya's-Commits

shaniyaclement committed:
Merge pull request #28 from blpham/hook_tests

Test hook for running data pipeline through API connection

shaniyaclement committed:
Merge branch 'main' into Shaniya's-Commits

Commit:
removed an overlooked conflict from when I merged the run test

Commit:
formatting all files created for both operators and hooks

shaniyaclement committed:
Merge pull request #23 from shaniyaclement/Shaniya's-Commits

PR For Host to Review Errors Discussed In Chat

shaniyaclement committed:
Merge pull request #29 from apache/main

Updating main to Apache
